### PR TITLE
Cumulus-642 Cumulus API uses Launchpad 

### PR DESCRIPTION
**Pull request by kkelly51**
Addresses [CUMULUS-642: Cumulus API uses Launchpad authentication](https://bugs.earthdata.nasa.gov/browse/CUMULUS-642)

### CODE REVIEW

File: docs/deployment/config_descriptions.md [hunk 1]
```diff
@@ -90,7 +90,7 @@ This table describes the fields that must be present in `config.yml` to successf
 | launchpad.api  | | launchpad API endpoint
 | launchpad.certificate  | launchpad.pfx | the Launchpad PKI certificate pfx file
 | launchpad.passphrase  | | the passphrase of the Launchpad PIK certificate
-| oauth.provider | earthdata | the authentication provider used for Cumulus API, possible values are 'earthdata', 'google', 'launchpad'
+| oauthProvider | earthdata | the authentication provider used for Cumulus API, possible values are 'earthdata', 'google', 'launchpad'
```
Comment (jennyhliu) [inline docs/deployment/config_descriptions.md]: oauth.provider has been used in backend_api_lambdas.yml.  We should keep this configuration, and may add another parameter inside, e.g. with default value
oauth:
    provider: earthdata
    userGroup: N/A

File: packages/api/app/launchpadAuth.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    }
+  } catch (error) {
+    if (error instanceof RecordDoesNotExist) {
+      const config = {
+        api: process.env.launchpad_api,
+        passphrase: process.env.launchpad_passphrase,
+        certificate: process.env.launchpad_certificate
+      };
+
+      const launchpadToken = new LaunchpadToken(config);
```
Comment (jennyhliu) [inline packages/api/app/launchpadAuth.js]: We could also add a method to @cumulus/common/launchpad to validate the token and user

File: packages/api/app/launchpadAuth.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
+
+'use strict';
+
+const LaunchpadToken = require('@cumulus/common/LaunchpadToken');
+const { RecordDoesNotExist } = require('@cumulus/common/errors');
+const { AccessToken } = require('../models');
+
+function checkUserGroups(userGroups) {
+  const cumulusGroup = process.env.cumulusUserGroup;
```
Comment (jennyhliu) [inline packages/api/app/launchpadAuth.js]: envs needs to be added to api/config/backend_api_lambdas.yml

File: packages/api/app/routes.js [hunk 4]
```diff
@@ -32,6 +33,10 @@ if (process.env.FAKE_AUTH === 'true') {
   ensureAuthorized = token.ensureAuthorized;
 }
 
+if (process.env.oauthProvider === 'launchpad') {
```
Comment (jennyhliu) [inline packages/api/app/routes.js]: If the token has been authorized, it should be validate for 1? hour, and within the hour, cumulus shouldn't have to validate  this token against Launchpad again.  The validation against Launchpad is expensive.

Comment (jennyhliu) [reply packages/api/app/routes.js]: actually, I see the token is saved

File: docs/deployment/cumulus_api_launchpad_authentication.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
+---
+id: cumulus_api_launchpad_authentication
```
Comment (jennyhliu) [inline docs/deployment/cumulus_api_launchpad_authentication.md]: need to add this page to sidebars website/sidebars.json

File: example/app/config.yml [hunk 6]
```diff
@@ -470,6 +470,16 @@ kk:
...
     ssh: true
     keyPairName: kakelly2-cumulus-sandbox
+  cmr:
+    oauthProvider: launchpad
+    username: CmrUserNotUsed
+    password: CmrPasswordNotUsed
+  launchpad:
+    api: 'https://api.launchpad.nasa.gov/icam/api/sm/v1'
+    passphrase: '{{LAUNCHPAD_PASSPHRASE}}'
+  oauth:
```
Comment (jennyhliu) [inline example/app/config.yml]: we want to configure `cumulus-from-source:` deployment to use `launchpad` authentication

File: packages/api/app/launchpadAuth.js [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+      const config = {
+        api: process.env.launchpad_api,
+        passphrase: process.env.launchpad_passphrase,
+        certificate: process.env.launchpad_certificate
+      };
+
+      const userGroup = process.env.oauth_user_group;
+      const verifyResponse = await launchpad.validateLaunchpadToken(config, token, userGroup);
+
+      if (verifyResponse.status === 'success') {
```
Comment (jennyhliu) [inline packages/api/app/launchpadAuth.js]: have you tested that 1. token is saved to the database 2. token is valid for an hour and if the token is used within an hour, no validateLaunchpadToken is called 3. token is invalid after an hour

File: docs/deployment/cumulus_api_launchpad_authentication.md (hunk 5 above)
Comment (jennyhliu) [reply docs/deployment/cumulus_api_launchpad_authentication.md]: need to reference this page from upper level document

Comment (jennyhliu) [reply docs/deployment/cumulus_api_launchpad_authentication.md]: need to update https://github.com/nasa/cumulus/edit/master/docs/deployment/README.md
line 48, line 198 example

File: packages/api/app/launchpadAuth.js (hunk 7 above)
Comment (kkelly51) [reply packages/api/app/launchpadAuth.js]: 1. yes, 2&3. Not exactly but I can refine,

**Pull request merged**